internals/plan/plan.go

```diff
@@ -641,179 +644,277 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 		}
 	}
 
-	// Ensure fields in combined layers validate correctly (and set defaults).
-	for name, service := range combined.Services {
-		if service.Command == "" {
-			return nil, &FormatError{
-				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
+	// Set defaults where required.
+	for _, service := range combined.Services {
+		if !service.BackoffDelay.IsSet {
+			service.BackoffDelay.Value = defaultBackoffDelay
+		}
+		if !service.BackoffFactor.IsSet {
+			service.BackoffFactor.Value = defaultBackoffFactor
+		}
+		if !service.BackoffLimit.IsSet {
+			service.BackoffLimit.Value = defaultBackoffLimit
+		}
+	}
+
+	for _, check := range combined.Checks {
+		if !check.Period.IsSet {
+			check.Period.Value = defaultCheckPeriod
+		}
+		if !check.Timeout.IsSet {
+			check.Timeout.Value = defaultCheckTimeout
+		}
+		if check.Timeout.Value > check.Period.Value {
+			// The effective timeout will be the period, so make that clear.
+			// `.IsSet` remains false so that the capped value does not appear
+			// in the combined plan output - and it's not *user* set - the
+			// effective default timeout is the minimum of (check.Period.Value,
+			// default timeout).
+			check.Timeout.Value = check.Period.Value
+		}
+		if check.Threshold == 0 {
+			// Default number of failures in a row before check triggers
+			// action, default is >1 to avoid flapping due to glitches. For
+			// what it's worth, Kubernetes probes uses a default of 3 too.
+			check.Threshold = defaultCheckThreshold
+		}
+	}
+
+	return combined, nil
+}
+
+// Validate checks that the layer is valid. It returns nil if all the checks pass, or
+// an error if there are validation errors.
+// See also Plan.Validate, which does additional checks based on the combined
+// layers.
+func (layer *Layer) Validate() error {
+	for name, service := range layer.Services {
+		if name == "" {
+			return &FormatError{
+				Message: fmt.Sprintf("cannot use empty string as service name"),
 			}
 		}
+		if name == "pebble" {
+			// Disallow service name "pebble" to avoid ambiguity (for example,
+			// in log output).
+			return &FormatError{
+				Message: fmt.Sprintf("cannot use reserved service name %q", name),
+			}
+		}
+		// Deprecated service names
+		if name == "all" || name == "default" || name == "none" {
+			logger.Noticef("Using keyword %q as a service name is deprecated", name)
+		}
+		if strings.HasPrefix(name, "-") {
+			return &FormatError{
+				Message: fmt.Sprintf(`cannot use service name %q: starting with "-" not allowed`, name),
+			}
+		}
+		if service == nil {
+			return &FormatError{
+				Message: fmt.Sprintf("service object cannot be null for service %q", name),
+			}
+		}
+		// ParseCommand needs the name to be set in order to generate a nice error message.
+		service.Name = name
```

Comment:
I'm a bit wary about having this in the `Validate` function, which I don't think should mutate anything. Why isn't the service name set at this point? It looks like `ParseLayer` is already looping through and setting the `Name` field.